File: LeetCode/Easy/9-add-two-numbers-Linked-List.py
# Definition for singly-linked list.
class ListNode():
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution():
    def addTwoNumbers(self, l1, l2):
        """Adds two numbers represented as linked lists."""
        
        dummy_node = ListNode()  # Dummy node to simplify edge cases
        tail = dummy_node  # Pointer to build the result list
        carry = 0  # Store carry from sum
        
        # Traverse both lists until we reach the end of both
        while l1 or l2 or carry:
            val1 = l1.val if l1 else 0  # Get value from l1 (0 if l1 is None)
            val2 = l2.val if l2 else 0  # Get value from l2 (0 if l2 is None)
            
            # Compute sum of digits + carry
            total = val1 + val2 + carry
            carry = total // 10  # Compute carry for next iteration | used to extract the first digit for ex val1 = 4 val2 = 6 = 10 | carry = 1
            sum_digit = total % 10  # Get the last digit|if the last digit is 9 then sum_diget will be = 9 

            print(f"Adding: {val1} + {val2} + carry({carry}) = {total} (sum digit: {sum_digit}, new carry: {carry})")
            
            # Append the new digit to the linked list
            tail.next = ListNode(sum_digit)
            tail = tail.next  # Move tail forward
           
            
            # Move l1 and l2 to next node (if exists)
            if l1: 
                l1 = l1.next
            if l2:
                l2 = l2.next
        
       # we return the result list skipping the dummy node which was used to simplify edge cases it was 0
                """tail = dummy_node means tail points to the same object as dummy_node.
        ✅ As we build the linked list, tail moves forward, but dummy_node always stays at the start.
        ✅ The final result starts from dummy_node.next, skipping the dummy placeholder"""
        return dummy_node.next  # Return the result list (skip dummy node)

# Helper function to create a linked list from a Python list
def create_linked_list(values):
    """Converts a list into a linked list."""
    if not values:
        return None
    head = ListNode(values[0])
    current = head
    for value in values[1:]:   
        current.next = ListNode(value)
        current = current.next
    return head

# Helper function to print a linked list
def print_linked_list(head):
    """Prints a linked list in a readable format."""
    values = []
    while head:
        values.append(head.val)
        head = head.next
    print(" -> ".join(map(str, values)))

# Create linked lists from given Python lists
l1 = create_linked_list([2,4,3])  # Represents 342
l2 = create_linked_list([5,6,4])  # Represents 465

# Print the input lists
print("List 1:")
print_linked_list(l1)

print("List 2:")
print_linked_list(l2)

# Add the two numbers
solution = Solution()
result = solution.addTwoNumbers(l1, l2)

# Print the result list
print("Sum List:")
print_linked_list(result)

File: LeetCode/Easy/9-add-two-numbers-Linked-List.ts
// Define the ListNode class
class ListNode {
    val: number;
    next: ListNode | null;
    constructor(val?: number, next?: ListNode | null) {
        this.val = (val === undefined ? 0 : val);
        this.next = (next === undefined ? null : next);
    }
}

// Define the Solution class
class Solution {
    addTwoNumbers(l1: ListNode | null, l2: ListNode | null): ListNode | null {
        let dummy = new ListNode(0); // Dummy node to simplify the logic
        let temp = dummy; // Temporary pointer to build the result
        let carry = 0; // Carry-over for addition

        // Iterate through both linked lists
        while (l1 !== null || l2 !== null || carry !== 0) {
            // Get the values of the current nodes (or 0 if the node is null)
            let val1 = l1 ? l1.val : 0;
            let val2 = l2 ? l2.val : 0;

            // Calculate the sum and the new carry
            let sum = val1 + val2 + carry;
            carry = Math.floor(sum / 10); // Update carry
            temp.next = new ListNode(sum % 10); // Create a new node for the result
            temp = temp.next; // Move the temp pointer

            // Move to the next nodes in the input lists
            if (l1 !== null) l1 = l1.next;
            if (l2 !== null) l2 = l2.next;
        }

        // Return the result (skip the dummy node)
        return dummy.next;
    }
}

// Helper function to convert an array to a linked list
function arrayToList(arr: number[]): ListNode | null {
    let dummy = new ListNode(0);
    let temp = dummy;
    for (let num of arr) {
        temp.next = new ListNode(num);
        temp = temp.next;
    }
    return dummy.next;
}

// Helper function to convert a linked list to an array
function listToArray(head: ListNode | null): number[] {
    let result: number[] = [];
    while (head !== null) {
        result.push(head.val);
        head = head.next;
    }
    return result;
}

// Example usage
const l1 = arrayToList([2, 4, 3]); // Convert array to linked list
const l2 = arrayToList([5, 6, 4]); // Convert array to linked list

const solution = new Solution();
const result = solution.addTwoNumbers(l1, l2); // Add the two linked lists

console.log(listToArray(result)); // Convert the result linked list to an array and print